### ISSUE TITLE
New MVT driver files sometimes fail Tangram draw

### ISSUE THREAD

**Issue by nvkelso** [tilezen/vector-datasource#1594]
The vast majority of new MVT files generated with Wagyu and VTZero (wrapped in https://github.com/tilezen/coanacatl) are rendering fine in Tangram. But I'm seeing a noticeable % that Tangram reports as having `unknown command X` where X is a value of 0, 5, and 6 so far.

Let's investigate if there's a different config we need to be using, or Tangram needs to learn how to read this different MVT variant. 

1. **unknown command 0** (outside of Paris)
    - tilezen/vector/v1/512/all/9/258/176.mvt
    - [176.mvt.zip](https://github.com/tilezen/vector-datasource/files/2349857/176.mvt.zip)
    - ![image](https://user-images.githubusercontent.com/853051/45052229-14017900-b03b-11e8-81db-876d5fd552e9.png)

2. **unknown command 3** (outside of Paris)
    - tilezen/vector/v1/512/all/10/517/352.mvt
    - [352.mvt.zip](https://github.com/tilezen/vector-datasource/files/2349892/352.mvt.zip)
    - <img width="715" alt="screen shot 2018-09-04 at 12 15 19" src="https://user-images.githubusercontent.com/853051/45052643-59727600-b03c-11e8-819d-539189262973.png">

2. **unknown command 4** (outside of Paris)
    - tilezen/vector/v1/512/all/14/8293/5634.mvt
    - [5634.mvt.zip](https://github.com/tilezen/vector-datasource/files/2349916/5634.mvt.zip)
    - ![image](https://user-images.githubusercontent.com/853051/45053109-dc480080-b03d-11e8-9941-21e0aea7c493.png)

2. **unknown command 5** (in Spain)
    - tilezen/vector/v1/512/all/8/127/95.mvt
    - [95.mvt.zip](https://github.com/tilezen/vector-datasource/files/2349858/95.mvt.zip)
    - ![image](https://user-images.githubusercontent.com/853051/45052265-2c719380-b03b-11e8-8cef-ecc8ede7de8c.png)

3.  **unknown command 6** (outside NYC)
    - tilezen/vector/v1/512/all/10/300/384.mvt
    - [384.mvt.zip](https://github.com/tilezen/vector-datasource/files/2349856/384.mvt.zip)
    - ![image](https://user-images.githubusercontent.com/853051/45052251-267bb280-b03b-11e8-8a63-4d8211a2c2f0.png)

**Comment by nvkelso** [comment]
/cc @bcamper as I'm seeing Tangram JS draw errors for the above MVT files. We're trying out Wagyu and VectorTileZero instead of Mapzen's old mvt python library for this, which seems to work the vast majority of times.

The Tangram errors I'm seeing are like:

```
blob:http://localhost:8000/9a3a3e90-4232-4306-b87b-b8b05b03a39f:1 Uncaught (in promise) Error: unknown command 0
    at n.loadGeometry (blob:http://localhost:8000/9a3a3e90-4232-4306-b87b-b8b05b03a39f:1)
    at t.value (blob:http://localhost:8000/9a3a3e90-4232-4306-b87b-b8b05b03a39f:14)
    at t.value (blob:http://localhost:8000/9a3a3e90-4232-4306-b87b-b8b05b03a39f:14)
    at blob:http://localhost:8000/9a3a3e90-4232-4306-b87b-b8b05b03a39f:14
```

**Comment by bcamper** [comment]
We could try upgrading the MVT reader (maintained by Mapbox) on the JS side if there's a newer version in case there is some field/encoding we are not currently supporting. But interested in @zerebubuth's feedback too, e.g. are these tiles validating with another tool so we can narrow it down?

**Comment by nvkelso** [comment]
Good point on the fields, @zerebubuth did mention this morning that we needed to catch property names and ensure they had the right encoding, maybe this is a symptom of that?

**Comment by nvkelso** [comment]
That's field work was in https://github.com/tilezen/coanacatl/pull/5.

**Comment by rmarianski** [comment]
I'd also suggest trying another decoder to see if it can read it. We can even try the other python one to see what happens :)

**Comment by zerebubuth** [comment]
I instrumented the [Python mapbox-vector-tile decoder](https://github.com/tilezen/mapbox-vector-tile/blob/46629e478935954fc2dcbd8d17d005546b8d31a0/mapbox_vector_tile/decoder.py#L88) with

```python
assert cmd in (CMD_SEG_END, CMD_MOVE_TO, CMD_LINE_TO)
```

and managed to successfully decode all the tiles. I'm not 100% certain that this means the tiles are OK, just that they're not obviously broken. It's possible that there's some corner case and they're being decoded wrong. But I think it might be more likely that the tile is being corrupted at some point, and that the tiles stored in the `.zip` files above might not be exactly the same as the ones that Tangram had to decode.

I'll try and repro tomorrow and see if I can get to a root cause.

**Comment by zerebubuth** [comment]
I was wrong about the tile being corrupted - I'm able to download the exact same tile in the browser and on the command line, and both decode OK with the Python `mapbox-vector-tile` library.

Although I wasn't able to debug into the tile decoding JS - for some reason Chrome keeps resetting my breakpoints when they're in web workers - I ported the tile decoding logic to Python, where it works just fine. So it looks like something lower level than that.

The issue seems to be that `coanacatl` is outputting an empty geometry, which is [not allowed](https://github.com/mapbox/vector-tile-spec/tree/master/2.1#42-features). The absence of a geometry causes the JS decoder to try and decode the beginning of the tile as a geometry, which will generally result in an "unknown command" sooner or later.

`coanacatl` has some protection against empty geometries, as Shapely generally returns a GeometryCollection for their type, which is unsupported and will raise an exception. And [the feature in question](https://www.openstreetmap.org/relation/6167088#map=9/48.5811/2.3387) seems like it might be large enough to exist in the tile.

I'll dig deeper to get a reproducible test case.

**Comment by bcamper** [comment]
Ugh yes breakpoints in the Tangram web workers are finicky. The native-ES6-bundled branch I am working on should address some of that (while introducing one or two different quirks of its own!).

Tangram is [currently using](https://github.com/tangrams/tangram/blob/master/package.json) `@mapbox/vector-tile` 1.3.0, with `pbf` 3.1.0. The [current version](https://github.com/mapbox/vector-tile-js/blob/master/package.json) of `@mapbox/vector-tile` appears to be 1.3.1 with `pbf` `^1.3.2` (??).

I would recommend testing against those versions (either standalone JS script test and/or Tangram branch).

**Comment by zerebubuth** [comment]
Fix to not generate empty geometries: https://github.com/tilezen/coanacatl/pull/6

> I would recommend testing against those versions (either standalone JS script test and/or Tangram branch).

I tested against both with a stand-alone program (thank you, `console.log`!), and it does the same thing. This could be fixed with a check after [parsing the feature tags](https://github.com/mapbox/vector-tile-js/blob/58df1e9344ee64f26deee84a9f54cee11fb95ef6/lib/vectortilefeature.js#L19) that `self._geometry` no longer has [the initial value of `-1`](https://github.com/mapbox/vector-tile-js/blob/58df1e9344ee64f26deee84a9f54cee11fb95ef6/lib/vectortilefeature.js#L15). However, given that missing geometry is not valid, one could make the argument that a check would be superfluous.

**Comment by nvkelso** [comment]
Additional followup in https://github.com/tilezen/coanacatl/pull/7.

**Comment by zerebubuth** [comment]
I _suspect_ the `Cannot read property 'push' of undefined` error has the same root cause as the `unknown command` errors. What's happening is that the MVT decoder is decoding a `lineTo` command, building up a list of vertices. However, when the code sees a `lineTo` command before any `moveTo` commands, it will try to `push` a point onto the end of an undefined variable. Under normal circumstances, with valid geometry, this would never happen.

In short, I think this is what happens when the geometry is empty and the unrelated bytes at the beginning of the tile encode a `lineTo` rather than an unknown command.

Frustratingly, I'd like to verify this but am unable to repro with the script I used last time. I can repro just fine in the browser, though. I thought the code versions were the same, but clearly I'm missing something.

**Comment by zerebubuth** [comment]
Thanks @bcamper! In the end I just copied `tangram.debug.js` locally and put in a few `console.log()`s. Looks like it was the same root cause as the `unknown command` errors. I just did a new render with the fix from https://github.com/tilezen/coanacatl/pull/7 and everything I've looked at so far seems OK.

**Comment by nvkelso** [comment]
Everything is working now. Closing!
